src/main/java/com/android/volley/AsyncRequestQueue.java

```diff
@@ -0,0 +1,581 @@
+                        // ordered
+                        // by the underlying Request.
+                        if (r1 instanceof RequestTask) {
+                            if (r2 instanceof RequestTask) {
+                                return ((RequestTask<?>) r1).compareTo(((RequestTask<?>) r2));
+                            }
+                            return 1;
+                        }
+                        return r2 instanceof RequestTask ? -1 : 0;
+                    }
+                });
+    }
+
+    /**
+     * Builder is used to build an instance of {@link AsyncRequestQueue} from values configured by
+     * the setters.
+     */
+    public static class Builder {
+        private AsyncCache mAsyncCache;
+        private final AsyncNetwork mNetwork;
+        private Cache mCache;
+        private ExecutorFactory mExecutorFactory;
+        private ResponseDelivery mResponseDelivery;
+
+        public Builder(AsyncNetwork asyncNetwork) {
+            if (asyncNetwork == null) {
+                throw new IllegalArgumentException("Network cannot be null");
+            }
+            mNetwork = asyncNetwork;
+            mExecutorFactory = null;
+            mResponseDelivery = new ExecutorDelivery(new Handler(Looper.getMainLooper()));
+            mAsyncCache = null;
+            mCache = null;
+        }
+
+        /**
+         * Sets the executor factory to be used by the AsyncRequestQueue. If this is not called,
+         * Volley will create suitable private thread pools.
+         */
+        public Builder setExecutorFactory(ExecutorFactory executorFactory) {
+            mExecutorFactory = executorFactory;
+            return this;
+        }
+
+        /**
+         * Sets the response deliver to be used by the AsyncRequestQueue. If this is not called, we
+         * will default to creating a new {@link ExecutorDelivery} with the application's main
+         * thread.
+         */
+        public Builder setResponseDelivery(ResponseDelivery responseDelivery) {
+            mResponseDelivery = responseDelivery;
+            return this;
+        }
+
+        /** Sets the AsyncCache to be used by the AsyncRequestQueue. */
+        public Builder setAsyncCache(AsyncCache asyncCache) {
+            mAsyncCache = asyncCache;
+            return this;
+        }
+
+        /** Sets the Cache to be used by the AsyncRequestQueue. */
+        public Builder setCache(Cache cache) {
+            mCache = cache;
+            return this;
+        }
+
+        /** Provides a default ExecutorFactory to use, if one is never set. */
+        private ExecutorFactory getDefaultExecutorFactory() {
+            return new ExecutorFactory() {
+                @Override
+                public ExecutorService createNonBlockingExecutor(
+                        BlockingQueue<Runnable> taskQueue) {
+                    return new ThreadPoolExecutor(
```

Comment:
nit: you could reduce some code duplication here with a helper method newThreadPoolExecutor that takes the maximum pool size and thread name suffix